tests/unit/module_utils/test_resource.py

```diff
@@ -0,0 +1,241 @@
+
+
+@pytest.mark.parametrize(
+    "node,expected",
+    [
+        ({"a": "resource:s1.a"}, {"a": "s1_value_a"}),
+        ({"a": "resource:s1.a", "b": "resource:s2.b"}, {'a': 's1_value_a', 'b': 's2_value_b'}),
+        ({"c": ["resource:s1.a", "resource:s2.b"]}, {'c': ['s1_value_a', 's2_value_b']}),
+        ({"a": "resource:s1.c"}, None),
+    ]
+)
+@patch(PATCH_BASE_PATH + "replace_reference")
+def test_resolve_refs(m_replace_reference, node, expected):
+
+    context = {
+        "s1": {"a": "s1_value_a", "b": "s1_value_b"},
+        "s2": {"a": "s2_value_a", "b": "s2_value_b"},
+    }
+
+    def replace_ref(ctxt, match):
+        tmp = ctxt
+        for k in match.group(1).split("."):
+            tmp = tmp[k]
+        return tmp
+
+    m_replace_reference.side_effect = replace_ref
+    check_mode = False
+
+    if expected is None:
+        with pytest.raises(KeyError):
+            resolve_refs(node, context, check_mode=check_mode)
+    else:
+        result = resolve_refs(node=node, context=context, check_mode=check_mode)
+        print(f"Result: {result} - Expected: {expected}")
+        assert result == expected
+
+
+@patch(PATCH_BASE_PATH + "replace_reference")
+def test_resolve_refs_with_check_mode(m_replace_reference):
+
+    context = {}
+
+    def raise_key_error():
+        return KeyError("missing key from dictionnary")
+
+    m_replace_reference.side_effect = raise_key_error()
+    check_mode = True
+
+    node = {"a": "resource:s1.a"}
+    result = resolve_refs(node=node, context=context, check_mode=check_mode)
+    assert result == node
+
+
+@pytest.fixture()
+def cloudclient():
+
+    class TestCloudClient(CloudClient):
+
+        def present(self, resource: Dict) -> Dict:
+            pass
+
+        def absent(self, resource: Dict) -> Dict:
+            pass
+
+    client = TestCloudClient()
+    client.has_pyyaml = MagicMock()
+    client.has_pyyaml.return_value = True
+
+    return client
+
+
+@pytest.mark.parametrize("state", ["present", "absent"])
+@pytest.mark.parametrize(
+    "desired_state,expected",
+    [
+        (
+            {
+                "r1": {"name": "resource:r2.r1_name"},
+                "r2": {"r1_name": "value1", "name": "value2"}
+            },
+            ["r2", "r1"]
+        ),
+        (
+            {
+                "r1": {"name": "resource:r2.r1_name"},
+                "r2": {"r1_name": "value1", "name": "resource:r3.r2_name"},
+                "r3": {"r2_name": "value2"}
+            },
+            ["r3", "r2", "r1"]
+        ),
+    ]
+)
+def test_cloudclient_sort_resources(cloudclient, state, desired_state, expected):
+
+    sorter = cloudclient.sort_resources(desired_state, state)
+    result = []
+    while sorter:
+        for name in sorter.get_ready():
+            result.append(name)
+            sorter.done(name)
+    if state == "absent":
+        expected.reverse()
+    print(f"Expected: {expected} - Result: {result}")
```

Comment:
Is this left over from debugging? Can it be removed?

Reply:
will be removed